[EAV] Improving the EAV attribute code validation, by not allowing to use n…

eduard13 opened the pull request:
### Description (*)
This PR improves the attribute code validation, namely while creating the attribute programmatically. Before you were able to create (programmatically) an attribute with not allowed chars.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#20943: No complete validation while creation of attributes.
2. magento/magento2#20766: AttributeCode column name length validation throws wrong error message

### Manual testing scenarios (*)
1. Try to create a new attribute using dots, or with a first char being not letter
2. Try to create an attribute with code lenth less than 1 or more than 30 characters

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Improving the eav attribute code validation, by not allowing to use not allowed chars

magento-engcom-team commented:
Hi @eduard13. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

rogyar commented on app/code/Magento/Eav/Model/Validator/Attribute/Code.php:
```diff
@@ -0,0 +1,63 @@
+<?php
+/**
```
It's a good idea to use strict types within a scope of new implementations

eduard13 replied:
Thank you, the strict typing was declared.
Also do you think it's a good idea to rename this variable name into a less meaningful name (in order to fix the failing pipeline)? 
https://github.com/magento/magento2/blob/adb1a8f6f16a0da2cada4d2a838328334b4e3a18/app/code/Magento/Eav/Model/Entity/Attribute.php#L134

rogyar replied:
You have implemented the naming in the same manner as other validators. So, it's ok.
You could use an alias in the `use` section for this class. Something like:
```php
use Magento\Eav\Model\Validator\Attribute\Code as AttributeCodeValidator
```

eduard13 replied:
@rogyar the alias was added.

Commit:
Declaring strict typing

Commit:
Using an alias for attribute code validator

Commit:
Adding returning strict typing

irajneeshgupta commented:
```
if (trim($attributeCode)
            && !preg_match('/^[a-z][a-z0-9_]*$/', trim($attributeCode))
        ) {

```
@eduard13 

https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Catalog/Controller/Adminhtml/Product/Attribute/Save.php#L200

I think validation pattern should be same for both admin and creation of attributes programmatically...if admin user can create such attributes , then why such limitations for developers hence i requested you to add it.

Thank you.

Commit:
Allowing usage of the capital letters in attribute code

irajneeshgupta commented:
@eduard13  :+1:

Commit:
Improving the validation rules on product attribute save

Commit:
Fixing the Regex expr

Commit:
Using the same regex rule as it is in attribute code validator

sivaschenko commented on app/code/Magento/Catalog/Controller/Adminhtml/Product/Attribute/Save.php:
```diff
@@ -196,8 +196,23 @@ public function execute()
+
                 $validatorAttrCode = new \Zend_Validate_Regex(
-                    ['pattern' => '/^[a-zA-Z\x{600}-\x{6FF}][a-zA-Z\x{600}-\x{6FF}_0-9]{0,30}$/u']
+                    ['pattern' => '/^[a-zA-Z]+[a-zA-Z0-9_]*$/u']
```
It would be nice to move this pattern to a constant, as it is used more than once (I think added validation class is a good place for such constant to be declared).

eduard13 replied:
Makes sense. 
Thanks.

sivaschenko commented on app/code/Magento/Eav/Model/Validator/Attribute/Code.php:
```diff
@@ -0,0 +1,66 @@
+     * @return bool
+     * @throws LocalizedException
+     */
+    public function isValid($attributeCode): bool
```
The method named `isValid` should return bool and should not throw an exception. Validation messages should be collected using `_addMessages` parent method. Appropriate exceptions should be thrown if needed by client classes (that can access messages via `getMessages` validator method)

sivaschenko commented on app/code/Magento/Eav/Model/Validator/Attribute/Code.php:
```diff
@@ -0,0 +1,66 @@
+ *
+ * Validation EAV attribute code
+ */
+class Code  extends AbstractValidator
```
```suggestion
class Code extends AbstractValidator
```

sivaschenko commented on app/code/Magento/Eav/Model/Validator/Attribute/Code.php:
```diff
@@ -0,0 +1,66 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\Eav\Model\Validator\Attribute;
```
Please cover the validator class with a unit test

eduard13 replied:
The class was covered by Unit Test, and also all the related tests (unit, integration) were updated.

sivaschenko reviewed (changes requested):
Hi @eduard13 thanks for the pull request! Please see my code review notes

Commit:
Merge branch '2.3-develop' into 2.3-develop-eav-attribute-code-validation

Commit:
Refactoring the validator usage. Updating the Unit/Integration tests. Covering the new validator class by Unit Test

eduard13 commented:
@sivaschenko thank you for the review, I've updated the implementation.
Please note that I've completely removed the validation from Save controller, and moved it into Validate one, as you shouldn't reach the Save if the validation is failed.

Could you please review the updates?
Thank you.

sivaschenko commented:
I believe this pull request also resolves https://github.com/magento/magento2/issues/20766

eduard13 commented:
@sivaschenko yes, it does.

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-4246](https://jira.corp.magento.com/browse/ENGCOM-4246) has been created to process this Pull Request

magento-engcom-team commented:
Hi @rogyar, thank you for the review.
[ENGCOM-4246](https://jira.corp.magento.com/browse/ENGCOM-4246) has been created to process this Pull Request

VasylShvorak commented:
✔️ QA passed

p-bystritsky committed:
Merge branch '2.3-develop' into 2.3-develop-eav-attribute-code-validation

p-bystritsky committed:
magento/magento2#20526: Absent ObjectManager fix.

p-bystritsky committed:
magento/magento2#20526: Static test fix.

eduard13 commented:
@magento-engcom-team is there any reason of marking the PR just as an Improvement? Not even a `Bug fix` or `Test Coverage`?
cc: @sidolov, @sivaschenko  
Thank you.

sidolov commented:
Hi @eduard13 , right now only members of the Community Engineering team are able to set up achievements for pull requests, that why our automated tool removed them. I reverted the changes because your pull request really complex and should be rewarded. 
Thank you!

eduard13 commented:
Hi @sidolov, glad to hear that, but whom should I ping for recalculating the reward points for March?
Thank you.

sidolov commented:
Hi @eduard13 , points already recalculated
